File: HeartForCharity/HeartForCharity.Model/Responses/DashboardResponse.cs
using System;
using System.Collections.Generic;

namespace HeartForCharity.Model.Responses
{
    public class DashboardResponse
    {
        public int ActiveCampaigns { get; set; }
        public int FinishedCampaigns { get; set; }
        public int TotalVolunteers { get; set; }
        public decimal TotalRaised { get; set; }
        public List<MonthlyDonationItem> MonthlyDonations { get; set; } = new();
        public List<DashboardReviewItem> RecentReviews { get; set; } = new();
        public List<CampaignProgressItem> CampaignProgress { get; set; } = new();
    }

    public class MonthlyDonationItem
    {
        public int Year { get; set; }
        public int Month { get; set; }
        public decimal Total { get; set; }
        public int Count { get; set; }
    }

    public class DashboardReviewItem
    {
        public string ReviewerName { get; set; } = null!;
        public string? ReviewerAvatarUrl { get; set; }
        public int Rating { get; set; }
        public string? Comment { get; set; }
        public DateTime CreatedAt { get; set; }
    }

    public class CampaignProgressItem
    {
        public int CampaignId { get; set; }
        public string Title { get; set; } = null!;
        public decimal TargetAmount { get; set; }
        public decimal CurrentAmount { get; set; }
    }
}

File: HeartForCharity/UI/heartforcharity_desktop/lib/model/responses/dashboard_response.dart
class DashboardResponse {
  final int activeCampaigns;
  final int finishedCampaigns;
  final int totalVolunteers;
  final double totalRaised;
  final List<MonthlyDonationItem> monthlyDonations;
  final List<DashboardReviewItem> recentReviews;
  final List<CampaignProgressItem> campaignProgress;

  DashboardResponse({
    required this.activeCampaigns,
    required this.finishedCampaigns,
    required this.totalVolunteers,
    required this.totalRaised,
    required this.monthlyDonations,
    required this.recentReviews,
    required this.campaignProgress,
  });

  factory DashboardResponse.fromJson(Map<String, dynamic> json) {
    return DashboardResponse(
      activeCampaigns: json['activeCampaigns'] as int,
      finishedCampaigns: json['finishedCampaigns'] as int,
      totalVolunteers: json['totalVolunteers'] as int,
      totalRaised: (json['totalRaised'] as num).toDouble(),
      monthlyDonations: (json['monthlyDonations'] as List)
          .map((e) => MonthlyDonationItem.fromJson(e))
          .toList(),
      recentReviews: (json['recentReviews'] as List)
          .map((e) => DashboardReviewItem.fromJson(e))
          .toList(),
      campaignProgress: (json['campaignProgress'] as List)
          .map((e) => CampaignProgressItem.fromJson(e))
          .toList(),
    );
  }
}

class MonthlyDonationItem {
  final int year;
  final int month;
  final double total;
  final int count;

  MonthlyDonationItem({
    required this.year,
    required this.month,
    required this.total,
    required this.count,
  });

  factory MonthlyDonationItem.fromJson(Map<String, dynamic> json) {
    return MonthlyDonationItem(
      year: json['year'] as int,
      month: json['month'] as int,
      total: (json['total'] as num).toDouble(),
      count: json['count'] as int,
    );
  }
}

class DashboardReviewItem {
  final String reviewerName;
  final String? reviewerAvatarUrl;
  final int rating;
  final String? comment;
  final DateTime createdAt;

  DashboardReviewItem({
    required this.reviewerName,
    this.reviewerAvatarUrl,
    required this.rating,
    this.comment,
    required this.createdAt,
  });

  factory DashboardReviewItem.fromJson(Map<String, dynamic> json) {
    return DashboardReviewItem(
      reviewerName: json['reviewerName'] as String,
      reviewerAvatarUrl: json['reviewerAvatarUrl'] as String?,
      rating: json['rating'] as int,
      comment: json['comment'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

class CampaignProgressItem {
  final int campaignId;
  final String title;
  final double targetAmount;
  final double currentAmount;

  CampaignProgressItem({
    required this.campaignId,
    required this.title,
    required this.targetAmount,
    required this.currentAmount,
  });

  double get progress => targetAmount > 0 ? (currentAmount / targetAmount).clamp(0.0, 1.0) : 0.0;

  factory CampaignProgressItem.fromJson(Map<String, dynamic> json) {
    return CampaignProgressItem(
      campaignId: json['campaignId'] as int,
      title: json['title'] as String,
      targetAmount: (json['targetAmount'] as num).toDouble(),
      currentAmount: (json['currentAmount'] as num).toDouble(),
    );
  }
}
